---
# assannphy - Bacterial assembly, annotation, and phylogenetics
---

### Short summary: 
This project aims to develop a Phython-based pipeline for assembling and comparing genomes from three bacterial strains, including Cyanobacteria, Actinobacteria, Firmicutes. Princple steps of the pipline are:

1.  Sequence identification and analysis
2. De novo genome assembly
3. Genome annotation
5. Phyogenetic analysis and comparison

### TA: Jessen

### GitHub Repo: 
[assannphy](https://github.com/bredeson/assannphy)

### Participants:   
 - Ankit Sharma
 - Chan Young Park
 - Jorge Antonio Tzec Interian

---
# GC and Amino Acid Content in the Whole Genome.
---

### Short summary: 
The main goals of the present project were to:
1. Write a script that takes a whole genome and/or the coding part of a genome as input(s) and outputs either DNA length, GC count, and GC content (whole genome) or DNA length, GC count, GC content, amino acid counts, and amino acid contents (coding genome). The outputs were formatted as a double dictionary, with the chromosome number as the key for the outer dictionary and the name of the operation (e.g., DNA length, GC count) as the key for the inner dictionary. The actual numbers were the elements of the inner dictionary. We ensured that the script could be run for different species with similar genomes and executed it for the human, blue whale, and mouse genomes for further analyses.
2. Present the numbers in a graphical way using Pandas, Matplotlib, and Jupyter Notebooks. The inputs were the dictionaries created in step 1. The outputs included: a) a bar graph showing GC content averaged across the whole genome for the whole and coding human genome, b) a bar graph showing GC content per chromosome for the whole and coding human genome, c) scatter plots showing the association between GC content (whole and coding human genome) and the content of each amino acid (coding human genome), and d) a bar graph comparing the GC content averaged across the whole genome for the whole and coding genomes between human, blue whale, and mouse.
3. Please note that we downloaded the genome data from Ensembl. A challenge that arose was that the whole genome was organized by chromosome, while the coding genome was organized by gene. Therefore, our additional goal 3) was to approximate the coding sequences per chromosome, test this approximation, and display it visually.

### TA: Maria Jose Blanco Salazar

### GitHub Repo: 
[Project GC](https://github.com/millce-cm/pfb2024_group_project_GC)

### Participants:   
 - Caitlin Millett 
 - Johanna Seitz-Holland
 - Rada Malko
 - Timothy Schwanitz


---
# LOLA-seq (snRNA-seq and ATAC-seq pipeline)
---

### Short summary: 
The goal of this project was to run a snRNA-seq and snATAC-seq analysis on a dataset which is derived from the cells of a human patient with ischemic cardiomyopathy, or heart failure. The analysis also involved a peak calling algorithm to generate files not provided by the authors. 

### TA: Robert Dalipovic

### GitHub Repo: 
[Project GC](https://github.com/rdalipo1/PFB-LOLA-seq)

### Participants:   
 - Nathan Harper
 - Rajendra Pangeni
 - Henry Arnold
 - Taylor Pio
 - Upendra Chalise


---
# Title
---

### Short summary: 

### TA: 

### GitHub Repo: 

### Participants:   
 - 
 - 
 - 
 - 


---
# Title
---

### Short summary: 

### TA: 

### GitHub Repo: 

### Participants:   
 - 
 - 
 - 
 - 


---
# Title
---

### Short summary: 

### TA: 

### GitHub Repo: 

### Participants:   
 - 
 - 
 - 
 - 


